include/rocksdb/trace_record_result.h

```diff
@@ -0,0 +1,145 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/status.h"
+#include "rocksdb/trace_record.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class MultiValuesTraceExecutionResult;
+class SingleValueTraceExecutionResult;
+class StatusOnlyTraceExecutionResult;
+
+// Base class for the results of all types of trace records.
+class TraceRecordResult {
+ public:
+  explicit TraceRecordResult(TraceType trace_type);
+
+  virtual ~TraceRecordResult();
+
+  // Trace type of the corresponding TraceRecord.
+  virtual TraceType GetTraceType() const;
+
+  class Handler {
+   public:
+    virtual ~Handler() {}
+
+    // Handle StatusOnlyTraceExecutionResult
+    virtual Status Handle(const StatusOnlyTraceExecutionResult& /*result*/) {
+      return Status::NotSupported(
+          "StatusOnlyTraceExecutionResult is not supported.");
+    }
+
+    // Handle SingleValueTraceExecutionResult
+    virtual Status Handle(const SingleValueTraceExecutionResult& /*result*/) {
+      return Status::NotSupported(
+          "SingleValueTraceExecutionResult is not supported.");
+    }
+
+    // Handle MultiValuesTraceExecutionResult
+    virtual Status Handle(const MultiValuesTraceExecutionResult& /*result*/) {
+      return Status::NotSupported(
+          "MultiValuesTraceExecutionResult is not supported.");
+    }
+  };
+
+  virtual Status Accept(Handler* handler) = 0;
+
+ private:
+  TraceType trace_type_;
+};
+
+// Base class for the execution results types of trace records.
+class TraceExecutionResult : public TraceRecordResult {
+ public:
+  TraceExecutionResult(uint64_t latency, TraceType trace_type);
+
+  virtual ~TraceExecutionResult() override;
+
+  // Request latency in nanoseconds.
+  virtual uint64_t GetLatency() const;
+
+ private:
+  uint64_t latency_;
```

Comment:
Sorry I forget, did we discuss what is the benefit of moving latency measurement to RocksDB? In any case it'd be good to write that in the description.

Reply:
Actually I feel it's better to report the start/end timestamp of the execution. Then the user can use these values to order the execution result for analysis, or compute latency, etc for other types of analysis.

Reply:
`NowNanos()` uses `CLOCK_MONOTONIC`, which is not real time:
https://linux.die.net/man/3/clock_gettime
And it may not be accurate between threads (cores): https://stackoverflow.com/questions/46893072/clock-gettimeclock-monotonic-monotonicity-across-cores-threads

Typically we cannot get current absolute time at nanos level, only to the Micros level.
So is it okay to use Micros? or switch back to latency?

Reply:
> `NowNanos()` uses `CLOCK_MONOTONIC`, which is not real time:
> https://linux.die.net/man/3/clock_gettime
> And it may not be accurate between threads (cores): https://stackoverflow.com/questions/46893072/clock-gettimeclock-monotonic-monotonicity-across-cores-threads
> 
> Typically we cannot get current absolute time at nanos level, only to the Micros level.
> So is it okay to use Micros? or switch back to latency?

Changed to `NowMicros()`. Is this OK?